test/Microsoft.ML.StaticPipelineTesting/StaticPipeTests.cs

```diff
@@ -0,0 +1,133 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using FakeStaticPipes;
+using Microsoft.ML.Data.StaticPipe;
+using Microsoft.ML.Data.StaticPipe.Runtime;
+using Microsoft.ML.Runtime;
+using Microsoft.ML.Runtime.Data;
+using Microsoft.ML.TestFramework;
+using System;
+using System.IO;
+using System.Text;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Microsoft.ML.StaticPipelineTesting
+{
+
+    public abstract class MakeConsoleWork : IDisposable
+    {
+        private readonly ITestOutputHelper _output;
+        private readonly TextWriter _originalOut;
+        private readonly TextWriter _textWriter;
+
+        public MakeConsoleWork(ITestOutputHelper output)
+        {
+            _output = output;
+            _originalOut = Console.Out;
+            _textWriter = new StringWriter();
+            Console.SetOut(_textWriter);
+        }
+
+        public void Dispose()
+        {
+            _output.WriteLine(_textWriter.ToString());
+            Console.SetOut(_originalOut);
+        }
+    }
+
+    public sealed class StaticPipeTests : MakeConsoleWork
+    {
+        public StaticPipeTests(ITestOutputHelper output)
+            : base(output)
+        {
+        }
+
+        [Fact]
+        public void SimpleTextLoaderCopyColumnsTest()
+        {
+            var env = new TlcEnvironment(new SysRandom(0), verbose: true);
+
+            const string data = "0 hello 3.14159 -0 2\n"
+                + "1 1 2 4 15";
+            var dataSource = new BytesStreamSource(data);
+
+            var text = TextLoader.CreateReader(env, ctx => (
```

Comment:
>text [](start = 16, length = 4)

Unless you want to retain the reference to the reader, I'd rather have 'text' be already data.

Reply:
Maybe. I think perhaps though this might be part of the consequence of separating model and data, unless you want to back off the ideas in and make data a possibly model bearing object again.

Reply:
We can think more about this. I feel like this is still in flux.

---
In reply to: [214165485](https://github.com/dotnet/machinelearning/pull/778#discussion_r214165485) [](ancestors = 214165485)